### ISSUE TITLE
Restrict login to users with a specific role

### ISSUE THREAD

**Issue by robintw** [djpugh/fastapi_aad_auth#74]
I'm trying to restrict the users allowed to login to my application to just those who have a specific role set.

I have set my roles up in the Azure Portal, and the role (`Data`) is defined in the manifest for my app. If I add or remove that role to my user, then I can see it is correctly listed (or not listed) in the `/me` endpoint once I've logged in.

I have tried to set the `AAD_ROLES` variable in my .env file to list the role that I want all users to have. I've specified it like this:

```
AAD_ROLES="[\"Data\"]"
```

However, I still find that I am able to login even when I haven't been assigned that role.

Is this the correct way to require a specific role when logging in? Have I specified the role correctly in the `AAD_ROLES` env var? (I tried various ways, and most of them gave JSON errors, but this doesn't seem to give an error)

Any help would be much appreciated.

**Comment by robintw** [comment]
A quick update to this:

I'm trying to do this 'manually' now, by checking the `user.roles` string when I check authentication. However, I can't seem to get this to work when I am using a function that is decorated with `@auth_provider.auth_required()`. The `request.user` object seems to be some sort of Starlette User object, rather than the user_klass object that is configured within fastapi_aad_auth. That means the user object doesn't have a `roles` attribute, and so I can't check that.

Does anyone have any suggestions as to how to get access to the full user object in `request.user`?

**Comment by djpugh** [comment]
Hi @robintw  - I'm just travelling at the moment, I do recall working through this, so let me check and update here (and in the docs)

**Comment by robintw** [comment]
Thanks @djpugh.

I worked out it was available in the auth state, but I couldn't work out how to get access to the auth state when I'm using the `@auth_provider.auth_required()` decorator, rather than the `Depends(...)` parameter. Any ideas?

Thanks so much for working on the other two things: both of those will really help me!

**Comment by djpugh** [comment]
Are you using this interactively rather than via API?

That method is currently only checking scopes, not roles or groups, I will add that logic in as well ;)

**Comment by djpugh** [comment]
You could in the short term create a custom user class that adds a method like:
```
from fastapi_aad_auth._base.state import User


class RoleUser(User):

    @property
    def permissions(self):
        """User Permissions."""
        permissions = super().permissions
        if self.roles:
            for role in self.roles:
                    permissions.append(role)
        return permissions[:]
```
You should then be able to do ``auth_required(scopes='authenticated <ROLE>')``

**Comment by djpugh** [comment]
@robintw - I've had a go at those fixes in #80 - unfortunately I don't have the parameters for my test instance set up on this laptop, so can't fully test it - I will try and do it this weekend when I'm back from my travels.
